File: Windows/Shiba.Shared/Singleton.cs
﻿using System;
using System.Collections.Concurrent;

namespace Shiba.Internal
{
    internal static class Singleton<T> where T : new()
    {
        private static readonly ConcurrentDictionary<Type, T> Instances = new ConcurrentDictionary<Type, T>();

        public static T Instance => Instances.GetOrAdd(typeof(T), t => new T());
    }
}

File: Android/shiba/src/main/java/moe/tlaster/shiba/common/Singleton.kt
package moe.tlaster.shiba.common

import java.util.concurrent.ConcurrentHashMap

internal object Singleton {
    val Instances = ConcurrentHashMap<Class<*>, Any>()
    inline fun <reified T> get() : T =
            Instances.getOrPut(T::class.java) {T::class.java.newInstance()} as T
} 